src/transforms/apply-operation.js

```diff
@@ -120,14 +132,46 @@ function insertText(state, operation) {
  *
  * @param {State} state
  * @param {Object} operation
+ *   @param {Boolean} operation.deep (optional) Join recursively the
+ *   respective last node and first node of the nodes' children. Like a zipper :)
  * @return {State}
  */
 
 function joinNode(state, operation) {
-  const { path, withPath } = operation
-  let { document } = state
-  document = document.joinNode(path, withPath)
-  state = state.merge({ document })
+  const { path, withPath, deep = false } = operation
```

Comment:
We need it to implement correctly the inverse of the `splitNode` operation.

Here is a case, as tested in `fixtures/on-history/undo/split-node-block`

```yaml
nodes:
  - kind: block
    type: paragraph
    nodes:
      - kind: text
        text: SplitHere
```

We `splitNode` the paragraph, at an `offset` of 3, we get this:

```yaml
nodes:
  - kind: block
    type: paragraph
    nodes:
      - kind: text
        text: Split
  - kind: block
    type: paragraph
    nodes:
      - kind: text
        text: Here
```

Now, to revert that, we need to `joinNode` both paragraph.

```yaml
nodes:
  - kind: block
    type: paragraph
    nodes:
      - kind: text
        text: Split
      - kind: text
        text: Here
```

And then `joinNode` again the children text:

```yaml
nodes:
  - kind: block
    type: paragraph
    nodes:
      - kind: text
        text: SplitHere

```

`splitNode` splits all the way down, we need a way to join all the way down too :). The `deep` option of `joinNode` does that, by joining nodes like a zipper. 

By default, `deep=false`, and `joinNode` only join two nodes, no further, as before.